src/query/interpreter.cpp

```diff
@@ -6095,35 +6127,47 @@ void Interpreter::Abort() {
     }
   });
 
-  bool decrement = true;
-
   // System tx
   // TODO Implement system transaction scope and the ability to abort
   system_transaction_.reset();
 
   // Data tx
-  auto expected = TransactionStatus::ACTIVE;
+  auto expected = transaction_status_.load();
+
+  // Transition from (ACTIVE | IDLE | TERMINATED) -> STARTED_ROLLBACK.
+  while (expected != TransactionStatus::ACTIVE && expected != TransactionStatus::IDLE &&
+         expected != TransactionStatus::TERMINATED) {
+    expected = transaction_status_.load();
+    std::this_thread::sleep_for(std::chrono::microseconds(1));
+  }
+
   while (!transaction_status_.compare_exchange_weak(expected, TransactionStatus::STARTED_ROLLBACK)) {
-    if (expected == TransactionStatus::TERMINATED || expected == TransactionStatus::IDLE) {
-      transaction_status_.store(TransactionStatus::STARTED_ROLLBACK);
-      decrement = false;
-      break;
+    if (expected != TransactionStatus::ACTIVE && expected != TransactionStatus::IDLE &&
+        expected != TransactionStatus::TERMINATED) {
+      while (expected != TransactionStatus::ACTIVE && expected != TransactionStatus::IDLE &&
+             expected != TransactionStatus::TERMINATED) {
+        expected = transaction_status_.load();
+        std::this_thread::sleep_for(std::chrono::microseconds(1));
+      }
     }
-    expected = TransactionStatus::ACTIVE;
-    std::this_thread::sleep_for(std::chrono::milliseconds(1));
   }
 
-  utils::OnScopeExit clean_status(
-      [this]() { transaction_status_.store(TransactionStatus::IDLE, std::memory_order_release); });
+  utils::OnScopeExit const clean_status([this]() {
+    for (TransactionStatus expected = TransactionStatus::STARTED_ROLLBACK;
+         !transaction_status_.compare_exchange_weak(expected, TransactionStatus::UPDATING_TRANSACTION_INFO);
+         expected = TransactionStatus::STARTED_ROLLBACK) {
+    }
+
+    current_transaction_.reset();
+    metadata_ = std::nullopt;
+    transaction_status_.store(TransactionStatus::IDLE, std::memory_order_release);
```

Comment:
this assumes that the state doesn't get changed from the moment you updated transaction info to the moment you are going into IDLE state which is basically a lock mechanism